components/merkledb/src/validation.rs

```diff
@@ -1,37 +1,128 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 //! Validation helpers for index names.
 
-/// Validate index name.
-pub fn is_valid_index_name<S: AsRef<str>>(name: S) -> bool {
+use crate::access::{AccessError, AccessErrorKind};
+use crate::IndexAddress;
+
+/// Validates index name.
+pub fn is_valid_index_full_name<S: AsRef<str>>(name: S) -> bool {
+    name.as_ref()
+        .as_bytes()
+        .iter()
+        .all(|&c| is_allowed_index_name_char(c) || c == b'.')
+}
+
+/// Validates index name prefix, it shouldn't contain the dot.
+pub fn is_valid_index_name_component<S: AsRef<str>>(name: S) -> bool {
     name.as_ref()
         .as_bytes()
         .iter()
         .copied()
-        .all(is_allowed_latin1_char)
+        .all(is_allowed_index_name_char)
 }
 
-/// Check that latin1 character is allowed in index name.
+/// Checks that character is allowed in index name.
 /// Only these combination of symbols are allowed:
 ///
-/// `[0..9]`, `[a-z]`, `[A-Z]`, `_`, `-`, `.`
-pub fn is_allowed_latin1_char(c: u8) -> bool {
+/// `[0..9]`, `[a-z]`, `[A-Z]`, `_`, `-`
+pub fn is_allowed_index_name_char(c: u8) -> bool {
     match c {
-        48..=57   // 0..9
-        | 65..=90   // A..Z
-        | 97..=122  // a..z
-        | 45..=46   // -.
-        | 95        // _
-        => true,
+        b'0'..=b'9' | b'A'..=b'Z' | b'a'..=b'z' | b'-' | b'_' => true,
         _ => false,
     }
 }
 
-/// Calls the `is_valid_name` function with the given name and panics if it returns `false`.
-pub(crate) fn assert_valid_name<S: AsRef<str>>(name: S) {
-    if name.as_ref().is_empty() {
-        panic!("Index name must not be empty")
+// Allow because it's looks more readable.
+#[allow(clippy::if_not_else)]
+fn check_valid_name<F>(
+    addr: IndexAddress,
+    predicate: F,
+    allowed_chars: String,
+) -> Result<(), AccessError>
+where
+    F: Fn(String) -> bool,
+{
+    let name = addr.name.clone();
```

Comment:
Why is cloning necessary here? I'd rather that `name` be cloned exactly if it's needed (i.e., if it's returned within an error).